Explicit bibkeys in the XML

mbollmann opened the pull request:
This PR commits bibliography keys to the XML, following the discussion in #1236. This will ensure that bibliography keys remain consistent over time.

It requires (in the XML schema) that all papers have a `<bibkey>` entry. A new script, [`write_bibkeys_to_xml.py`](https://github.com/acl-org/acl-anthology/blob/480e8296101aca6b3d14606fc9dd4c13547e33bc/bin/write_bibkeys_to_xml.py), loads the Anthology and creates bibkeys for papers that are currently lacking them, using the same algorithm as before. For ingesting new papers, it's probably easiest to copy the core logic of that process ([just these two lines, really](https://github.com/acl-org/acl-anthology/blob/480e8296101aca6b3d14606fc9dd4c13547e33bc/bin/write_bibkeys_to_xml.py#L79-L81)) somewhere into the ingestion script; I'd leave that to @mjpost since he knows it best. :)

mjpost commented:
Oh, great, I was just thinking about this.

I haven't looked at the code yet. One drawback it seems to me is that ingestion now requires ensuring keys are unique, which in turn requires loading the Anthology class, which has a high overhead. How much work would it be to incorporate some aspects of #835 into this? Ingestion is probably our core operation, one that gets run and repeated all the time, and I would very much like not to encumber it with this load time.

Note that when hiring an assistant, one of the applicants discovered that if you left out the `importdir` argument to the `Anthology` class, it loaded faster, and you still got some of its functionality. Would this suffice? I ask in case someone has time to look into it.

mbollmann commented:
> I haven't looked at the code yet. One drawback it seems to me is that ingestion now requires ensuring keys are unique, which in turn requires loading the Anthology class, which has a high overhead. How much work would it be to incorporate some aspects of #835 into this? Ingestion is probably our core operation, one that gets run and repeated all the time, and I would very much like not to encumber it with this load time.

I haven't really found any promising solution for #835 yet. I suspect it will require some thorough profiling and restructuring of the code. If you feel it's important I can put it as the next item on my to-do list, but I wouldn't hold my breath for it for this PR.

> Note that when hiring an assistant, one of the applicants discovered that if you left out the `importdir` argument to the `Anthology` class, it loaded faster, and you still got some of its functionality. Would this suffice? I ask in case someone has time to look into it.

I don't see a single useful thing you could do with the class without loading the data files (which won't happen if you leave out `importdir`). Also, it's the nature of the beast that in order to guarantee uniqueness of bibkeys, you need to look at every single XML file we have.

Two things:

1. The `Anthology` class should actually load faster now since the bibkey creation code doesn't need to run anymore on instantiation. But the difference may be minimal, I haven't timed it yet.
2. It should be simple to create a small spin-off function for `ingest.py` that only grabs the bibkeys from the XML and creates the new ones. It would essentially duplicate a bit of code, which is not ideal, but if this is the only functionality from the `Anthology` class that we need there and you feel the speed-up would be worth it, we could go this route (until an awesome solution for #835 removes the need for this hack).

mbollmann commented:
Oh, I also made this button for copying the bibkey, which was requested as a feature to use in combination with the full Anthology bib file:
![grafik](https://user-images.githubusercontent.com/11348502/127069559-e1318191-aeb2-49f5-8be3-4c31768031cd.png)

akoehn commented on bin/anthology/index.py:
```diff
@@ -61,8 +61,9 @@ class AnthologyIndex:
     """Keeps an index of persons, their associated papers, paper bibliography
```
This is a good time to extend the docstring for this constructor

mjpost replied:
I agree, it would be helpful to have even roughly-written documentation for these classes; I often have to remind myself.

akoehn commented on data/xml/schema.rnc:
```diff
@@ -36,6 +36,7 @@ Paper = element paper {
   attribute ingest-date { xsd:date }?,
   (element abstract { MarkupText }?
    & attachment*
+   & element bibkey { text }
```
Maybe make it check against `[a-zA-Z0-9-]+` ? (same for second occurence below)

akoehn reviewed (commented):
Looks good, just two minor things.

mjpost commented:
Okay, I am going to try to add redundant code that allows us to get just the bibkeys, for ingestion time. I hope to do this sometime this week. Feel free to do it yourself if you find time. We should try to merge this asap since it affects so many files.

mbollmann commented:
I addressed all the comments, and merged in ACL 2021 so the PR build (hopefully) doesn't fail anymore.

@mjpost, I added code to `ingest.py` that loads only the bibkeys. It's ~735ms on my machine. I'd leave it to you to determine where in the process the new bibkeys should be created — currently, `AnthologyIndex.create_bibkey()` expects a `Paper` object to be passed which you never create during ingestion; but since `AnthologyIndex.create_bibkey()` is now no longer used internally within the library, I guess you could also change it to just take the relevant `Paper` attributes as arguments directly. Let me know if you need any help here.

mjpost commented:
Drat, Marcel, I’m sorry, I forgot that Xinru added DOIs to ACL, which creates a lot of merge conflicts. I’m going to approve this, and we can merge it as soon as these conflicts are resolved, so we don’t have to keep dealing with this.

mbollmann commented:
Pretty sure I know how to solve this quickly with some git magic, will try tomorrow.

And of course I'm happy to merge this, but you should probably make a note to update the ingestion script then :)

mjpost commented:
Man, do *not* click on the "Files" tab of this PR! It has hosed every machine I've accidentally done it on.

Yes, we can fix `ingest.py` on our next ingestion. We won't be able to miss it because the schema requires the key.

I'll leave merging to you.

mbollmann commented:
Hmm weird, nothing bad happens on my end browsing through the Files tab ... ¯\\_(ツ)_/¯

I'll merge later because I noticed the copy-to-clipboard button for the bibkey doesn't actually work, and I don't quite know why right now. Will look into it tomorrow.
